cacheops/transaction.py

```diff
@@ -73,13 +73,17 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_value, traceback):
         connection = get_connection(self.using)
-        self._no_monkey.__exit__(self, exc_type, exc_value, traceback)
-        if not connection.closed_in_transaction and exc_type is None and \
-                not connection.needs_rollback:
-            if transaction_states[self.using]:
-                transaction_states[self.using].commit()
-        else:
+        try:
+            self._no_monkey.__exit__(self, exc_type, exc_value, traceback)
+        except DatabaseError:
             transaction_states[self.using].rollback()
+        finally:
+            if not connection.closed_in_transaction and exc_type is None and \
+                    not connection.needs_rollback:
+                if transaction_states[self.using]:
+                    transaction_states[self.using].commit()
+            else:
+                transaction_states[self.using].rollback()
```

Comment:
It looks like rollback might be called twice. Also, both commit and rollback might be called. Changing `finally` to `else` will fix both of that.

Reply:
Thanks, fixed